### PR TITLE
add boundary faces option to mesh_dual

### PR DESCRIPTION
**Pull request by GeneKao**
@brgcode @Licini I have extended the `mesh_dual` function with an optional parameter to construction boundary faces. 
An optional parameter `boundary` is set to `0` by default (Can later have more options number for how we deal with boundary) without boundary faces making sure the compatibility with old code such as TNA form-finding usages.  

Screenshot 2022-04-28 at 14 29 50
Screenshot 2022-04-28 at 14 29 09

### What type of change is this?

- [X] Bug fix in a **backwards-compatible** manner.
- [X] New feature in a **backwards-compatible** manner.
- [ ] Breaking change: bug fix or new feature that involve incompatible API changes.
- [ ] Other (e.g. doc update, configuration, etc)

### Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [X] I added a line to the `CHANGELOG.md` file in the `Unreleased` section under the most fitting heading (e.g. `Added`, `Changed`, `Removed`).
- [X] I ran all tests on my computer and it's all green (i.e. `invoke test`).
- [X] I ran lint on my computer and there are no errors (i.e. `invoke lint`).
- [ ] I added new functions/classes and made them available on a second-level import, e.g. `compas.datastructures.Mesh`.
- [ ] I have added tests that prove my fix is effective or that my feature works.
- [X] I have added necessary documentation (if appropriate)

### CODE REVIEW

File: src/compas/datastructures/mesh/duality.py [hunk 1]
```diff
@@ -21,6 +22,9 @@ def mesh_dual(mesh, cls=None):
     cls : Type[:class:`~compas.datastructures.Mesh`], optional
         The type of the dual mesh.
         Defaults to the type of the provided mesh object.
+    boundary: float, optional
```
Comment (ricardoavelino) [inline src/compas/datastructures/mesh/duality.py]: should say `str` here

File: src/compas/datastructures/mesh/duality.py [hunk 2]
```diff
@@ -21,6 +22,9 @@ def mesh_dual(mesh, cls=None):
...
+        boundary mode for the dual mesh
+        Default mode is 0, not create faces on boundaries
```
Comment (ricardoavelino) [inline src/compas/datastructures/mesh/duality.py]: Boundary mode for the dual mesh
Default mode is 0, which will not add new faces to the boundaries

**Review by ricardoavelino — COMMENTED** [review]
Perhaps we can add a test for this function with `boundary=1` in the `/tests/` folder. Checking for an example diagram which is the total number of faces that it returns

**Comment by duchaoyu** [comment]
image
I was also testing the boundary blocks. I realise that if there's a triangle face in the corner, this will lead to a leaf in the dual(see picture). This leaf will be ignored by the current `mesh_dual` function. And this leaf is also tricky for the boundary block...

**Comment by GeneKao** [comment]
> I was also testing the boundary blocks. I realise that if there's a triangle face in the corner, this will lead to a leaf in the dual(see picture). This leaf will be ignored by the current mesh_dual function. And this leaf is also tricky for the boundary block...

@duchaoyu Can you send me the example mesh to reproduce your error? I think this can be easily fixed.

**Review by tomvanmele — COMMENTED** [review]
please revise using a variation of the following

```python
dual = mesh.dual()
dual.flip_cycles()

edge_vertex = {}
for u, v in mesh.edges_on_boundary():
    x, y, z = mesh.edge_midpoint(u, v)
    edge_vertex[u, v] = edge_vertex[v, u] = dual.add_vertex(x=x, y=y, z=z)

for vertex in mesh.vertices_on_boundary():
    vertices = []
    nbrs = mesh.vertex_neighbors(vertex, ordered=True)[::-1]
    vertices.append(edge_vertex[vertex, nbrs[0]])
    for nbr in nbrs[:-1]:
        vertices.append(mesh.halfedge_face(vertex, nbr))
    vertices.append(edge_vertex[vertex, nbrs[-1]])
    dual.add_face(vertices)
```

**Comment by duchaoyu** [comment]
> > I was also testing the boundary blocks. I realise that if there's a triangle face in the corner, this will lead to a leaf in the dual(see picture). This leaf will be ignored by the current mesh_dual function. And this leaf is also tricky for the boundary block...
> 
> @duchaoyu Can you send me the example mesh to reproduce your error? I think this can be easily fixed.

I think I have fixed the bug by modifying your code... Probably not the best solution for the moment. But now the dual mesh matches the boundary of the original mesh
https://colab.research.google.com/drive/1L9sYa1IEhn2-sbmxjz6aO7W_DKwcYYi2#scrollTo=vmM_F_-H3NA7
image
image

**Comment by tomvanmele** [comment]
Screenshot 2022-04-29 at 08 55 55

```python
from compas.geometry import Polygon
from compas_cgal.triangulation import refined_delaunay_mesh
from compas.datastructures import Mesh
from compas.datastructures import trimesh_split_edge
from compas_view2.app import App

poly = Polygon.from_sides_and_radius_xy(4, 10)

V, F = refined_delaunay_mesh(poly, maxlength=3.0)

mesh = Mesh.from_vertices_and_faces(V, F)

# preprocess the corner faces
for face in mesh.faces_on_boundary():
    if mesh.face_degree(face) > 1:
        continue
    for u, v in mesh.face_halfedges(face):
        if mesh.is_edge_on_boundary(v, u):
            continue
        trimesh_split_edge(mesh, u, v)
        break

dual: Mesh = mesh.dual()
dual.flip_cycles()

edge_vertex = {}
for u, v in mesh.edges_on_boundary():
    x, y, z = mesh.edge_midpoint(u, v)
    edge_vertex[u, v] = edge_vertex[v, u] = dual.add_vertex(x=x, y=y, z=z)

vertex_vertex = {}
for vertex in mesh.vertices_on_boundary():
    x, y, z = mesh.vertex_coordinates(vertex)
    vertex_vertex[vertex] = dual.add_vertex(x=x, y=y, z=z)

for vertex in mesh.vertices_on_boundary():
    vertices = [vertex_vertex[vertex]]
    nbrs = mesh.vertex_neighbors(vertex, ordered=True)[::-1]
    vertices.append(edge_vertex[vertex, nbrs[0]])
    for nbr in nbrs[:-1]:
        vertices.append(mesh.halfedge_face(vertex, nbr))
    vertices.append(edge_vertex[vertex, nbrs[-1]])
    dual.add_face(vertices)

viewer = App()
viewer.add(mesh, show_faces=False)
viewer.add(dual)

viewer.show()
```

**Comment by duchaoyu** [comment]
> Screenshot 2022-04-29 at 08 55 55
> 
> ```python
> from compas.geometry import Polygon
> from compas_cgal.triangulation import refined_delaunay_mesh
> from compas.datastructures import Mesh
> from compas.datastructures import trimesh_split_edge
> from compas_view2.app import App
> 
> poly = Polygon.from_sides_and_radius_xy(4, 10)
> 
> V, F = refined_delaunay_mesh(poly, maxlength=3.0)
> 
> mesh = Mesh.from_vertices_and_faces(V, F)
> 
> # preprocess the corner faces
> for face in mesh.faces_on_boundary():
>     if mesh.face_degree(face) > 1:
>         continue
>     for u, v in mesh.face_halfedges(face):
>         if mesh.is_edge_on_boundary(v, u):
>             continue
>         trimesh_split_edge(mesh, u, v)
>         break
> 
> dual: Mesh = mesh.dual()
> dual.flip_cycles()
> 
> edge_vertex = {}
> for u, v in mesh.edges_on_boundary():
>     x, y, z = mesh.edge_midpoint(u, v)
>     edge_vertex[u, v] = edge_vertex[v, u] = dual.add_vertex(x=x, y=y, z=z)
> 
> vertex_vertex = {}
> for vertex in mesh.vertices_on_boundary():
>     x, y, z = mesh.vertex_coordinates(vertex)
>     vertex_vertex[vertex] = dual.add_vertex(x=x, y=y, z=z)
> 
> for vertex in mesh.vertices_on_boundary():
>     vertices = [vertex_vertex[vertex]]
>     nbrs = mesh.vertex_neighbors(vertex, ordered=True)[::-1]
>     vertices.append(edge_vertex[vertex, nbrs[0]])
>     for nbr in nbrs[:-1]:
>         vertices.append(mesh.halfedge_face(vertex, nbr))
>     vertices.append(edge_vertex[vertex, nbrs[-1]])
>     dual.add_face(vertices)
> 
> viewer = App()
> viewer.add(mesh, show_faces=False)
> viewer.add(dual)
> 
> viewer.show()
> ```

👍‍👍‍👍‍ omg it looks so clean... preprocessing the corner faces is also a great idea!!

**Comment by juney-lee** [comment]
`vertices_on_biundary()` should also become `vertices_on_boundaries()` in mesh.dual()...

i.e. shells with holes

as currently written, you get this in interior boundaries.


image

File: src/compas/datastructures/mesh/duality.py [hunk 3]
```diff
@@ -24,24 +23,41 @@ def mesh_dual(mesh, cls=None, boundary=0):
         Defaults to the type of the provided mesh object.
     boundary: float, optional
         boundary mode for the dual mesh
-        Default mode is 0, not create faces on boundaries
+        Default mode is 0, not create faces on boundaries.
+        1, create faces on mesh edges, not include original mesh boundary vertices.
+        2. create faces on mesh edges and include original mesh boundary vertices on the corner.
+        3. create faces on mesh edges and include all original mesh boundary vertices.
```
Comment (tomvanmele) [inline src/compas/datastructures/mesh/duality.py]: in my opinion there are only three options

1. don't include the boundary
2. include the boundary without the original vertices
3. include the boundary with the original vertices

```python
def mesh_dual(mesh, cls=None, include_boundary=False, preserve_boundary_vertices=False):
```

File: src/compas/datastructures/mesh/duality.py [hunk 4]
```diff
@@ -24,24 +23,41 @@ def mesh_dual(mesh, cls=None, boundary=0):
...
+    Examples
+    --------
+    >>> from compas.datastructures import Mesh
+    >>> mesh = Mesh.from_obj(compas.get('faces.obj'))
+    >>> mesh.delete_face(11)
+    >>> mesh.delete_face(6)
+    >>> mesh.delete_face(7)
+    >>> mesh.quads_to_triangles()
+    >>> mesh = mesh.subdivide('corner')
+    >>> dual = mesh.dual(boundary=3)
```
Comment (tomvanmele) [inline src/compas/datastructures/mesh/duality.py]: `boundary=3` is completely meaningless...

File: src/compas/datastructures/mesh/duality.py [hunk 5]
```diff
@@ -24,24 +23,41 @@ def mesh_dual(mesh, cls=None, boundary=0):
...
+
     """
     if not cls:
         cls = type(mesh)
 
     dual = cls()
 
+    mesh.unify_cycles()
```
Comment (tomvanmele) [inline src/compas/datastructures/mesh/duality.py]: potentially expensive so should be left up to the caller

File: src/compas/datastructures/mesh/duality.py [hunk 6]
```diff
@@ -54,28 +70,53 @@ def mesh_dual(mesh, cls=None, boundary=0):
...
+                num_faces += 1
+            else:
+                boundary_fids.append(edge_vertex[vertex, nbr_vertex])
 
-                if geometric_key(pt) not in boundary_xyz and num_faces not in vertex_xyz:
-                    vertex_xyz[num_faces] = pt
-                    boundary_xyz[geometric_key(pt)] = num_faces
-                    num_faces += 1
+        if vertex in outer and len(faces) == 1 and (boundary == 2 or boundary == 3):
+            boundary_fids.insert(len(faces) + 1, current_face)
```
Comment (tomvanmele) [inline src/compas/datastructures/mesh/duality.py]: why are we making this so complicated?

**Comment by GeneKao** [comment]
> please revise using a variation of the following
> 
> ```python
> dual = mesh.dual()
> dual.flip_cycles()
> 
> edge_vertex = {}
> for u, v in mesh.edges_on_boundary():
>     x, y, z = mesh.edge_midpoint(u, v)
>     edge_vertex[u, v] = edge_vertex[v, u] = dual.add_vertex(x=x, y=y, z=z)
> 
> for vertex in mesh.vertices_on_boundary():
>     vertices = []
>     nbrs = mesh.vertex_neighbors(vertex, ordered=True)[::-1]
>     vertices.append(edge_vertex[vertex, nbrs[0]])
>     for nbr in nbrs[:-1]:
>         vertices.append(mesh.halfedge_face(vertex, nbr))
>     vertices.append(edge_vertex[vertex, nbrs[-1]])
>     dual.add_face(vertices)
> ```

there are bugs having duplicated vertices and faces
Screenshot 2022-05-02 at 11 08 38

so I improve the code in 5f9a19a base on your suggestion and without `geometric_key`. 

```python
import compas
from compas.geometry import Point
from compas.datastructures import Mesh
from compas_view2.app import App
from compas_view2.objects import Collection
from compas_view2.shapes import Text

mesh = Mesh.from_obj(compas.get('faces.obj'))

dual = mesh.dual(boundary=3)

view = App(width=1200, height=800, show_grid=False)
view.add(mesh, show_faces=False)
view.add(dual, opacity=0.4, facecolor=(0, 0, 1), linecolor=(1, 0, 0))

for i, v in enumerate(dual.vertices()):
    t = Text(str(v), dual.vertex_coordinates(v), height=30)
    view.add(t, color=(0, .5, 0))

for face in dual.faces():
    t = Text(str(face), dual.face_centroid(face), height=30)
    view.add(t, color=(0, 0, 1))

view.add(Collection([Point(*dual.vertex_coordinates(p)) for p in dual.vertices()]), color=(1, 0, 0))

view.run()
```
Screenshot 2022-05-02 at 11 18 31

File: src/compas/datastructures/mesh/duality.py (hunk 4 above)
Comment (GeneKao) [reply src/compas/datastructures/mesh/duality.py]: agree, I will change this.

File: src/compas/datastructures/mesh/duality.py (hunk 5 above)
Comment (GeneKao) [reply src/compas/datastructures/mesh/duality.py]: I will remove it.

File: src/compas/datastructures/mesh/duality.py (hunk 3 above)
Comment (GeneKao) [reply src/compas/datastructures/mesh/duality.py]: @brgcode So this one doesn't make sense? It only boundary vertices on the corner, not on edges. 
Screenshot 2022-05-02 at 11 32 32

Comment (tomvanmele) [reply src/compas/datastructures/mesh/duality.py]: i would indeed say it only makes sense because of the specific boundary conditions. but not sure. have to think about it...

**Comment by GeneKao** [comment]
@brgcode @juney-lee @duchaoyu @ricardoavelino This will also work with corner cases and many boundaries. 

Screenshot 2022-05-02 at 11 42 51

``` python
import compas
from compas.geometry import Point
from compas.datastructures import Mesh
from compas_view2.app import App
from compas_view2.objects import Collection
from compas_view2.shapes import Text

mesh = Mesh.from_obj(compas.get('faces.obj'))

mesh.delete_face(18)
mesh.delete_face(11)
mesh.delete_face(6)
mesh.delete_face(7)
mesh.quads_to_triangles()
mesh = mesh.subdivide('corner')

dual = mesh.dual(boundary=3)

view = App(width=1200, height=800, show_grid=False)
view.add(mesh, show_faces=False)
view.add(dual, opacity=0.4, facecolor=(0, 0, 1), linecolor=(1, 0, 0))

for i, v in enumerate(dual.vertices()):
    t = Text(str(v), dual.vertex_coordinates(v), height=30)
    view.add(t, color=(0, .5, 0))

for face in dual.faces():
    t = Text(str(face), dual.face_centroid(face), height=30)
    view.add(t, color=(0, 0, 1))

view.add(Collection([Point(*dual.vertex_coordinates(p)) for p in dual.vertices()]), color=(1, 0, 0))

view.run()
```

**Comment by tomvanmele** [comment]
just check if the first and last edge of the boundary belong to the same face to avoid the duplication

File: src/compas/datastructures/mesh/duality.py (hunk 3 above)
Comment (GeneKao) [reply src/compas/datastructures/mesh/duality.py]: ok, I just keep this option for now as `preserve_corner_vertices` in case someone needs it. 
now it's 
```python 
def mesh_dual(mesh, cls=None, include_boundary=False, preserve_boundary_vertices=False, preserve_corner_vertices=False):
```

Comment (ricardoavelino) [reply src/compas/datastructures/mesh/duality.py]: Wouldn't it be simpler to only have `preserve_boundary_vertices=True` and this would keep the corner vertices?

File: src/compas/datastructures/mesh/duality.py (hunk 6 above)
Comment (GeneKao) [reply src/compas/datastructures/mesh/duality.py]: I tried to make this dual mesh as general and simple as possible. For example, we have to consider some special cases when mesh face id is not starting from 0 and is a bit random.

**Comment by GeneKao** [comment]
So currently in commit 0423ec6 we have those options: 

```python 
dual = mesh.dual(include_boundary=False, preserve_boundary_vertices=False, preserve_corner_vertices=False)
```
Screenshot 2022-05-02 at 16 56 50

```python 
dual = mesh.dual(include_boundary=True, preserve_boundary_vertices=False, preserve_corner_vertices=False)
```
Screenshot 2022-05-02 at 16 58 26

```python 
dual = mesh.dual(include_boundary=True, preserve_boundary_vertices=True, preserve_corner_vertices=False)
```
Screenshot 2022-05-02 at 16 59 54

```python 
dual = mesh.dual(include_boundary=True, preserve_boundary_vertices=False, preserve_corner_vertices=True)
```

Screenshot 2022-05-02 at 17 00 59

File: src/compas/datastructures/mesh/duality.py (hunk 3 above)
Comment (GeneKao) [reply src/compas/datastructures/mesh/duality.py]: @ricardoavelino see my latest comments below.

Comment (ricardoavelino) [reply src/compas/datastructures/mesh/duality.py]: Just saw it. Yes, indeed your PR cover all cases

**Comment by tomvanmele** [comment]
a few things

1. the orientation of the dual should stay as before. otherwise code relying on it will break.
2. identifiers of the dual vertices should match the identifiers of the primal faces. otherwise code relying on this will break.
3. in my opinion, there are only 2 options: include the boundary, or not include the boundary. anything in between sounds like it relies on assumptions that may or may not be generally true. (superfluous boundary edges can be easily removed if needed in a post-processing step)

therefore i propose something like the following

```python
def mesh_dual(mesh, cls=None, include_boundary=False):
    """Construct the dual of a mesh.

    Parameters
    ----------
    mesh : :class:`~compas.datastructures.Mesh`
        A mesh object.
    cls : Type[:class:`~compas.datastructures.Mesh`], optional
        The type of the dual mesh.
        Defaults to the type of the provided mesh object.
    include_boundary : bool, optional
        If True, ...

    Returns
    -------
    :class:`~compas.datastructures.Mesh`
        The dual mesh object.

    """
    if not cls:
        cls = type(mesh)

    dual = cls()

    face_centroid = {face: mesh.face_centroid(face) for face in mesh.faces()}
    outer = set(flatten(mesh.vertices_on_boundaries()))
    inner = list(set(mesh.vertices()) - outer)
    vertex_xyz = {}
    face_vertices = {}

    for vertex in inner:
        faces = mesh.vertex_faces(vertex, ordered=True)
        for face in faces:
            if face not in vertex_xyz:
                vertex_xyz[face] = face_centroid[face]
        face_vertices[vertex] = faces

    for vertex in vertex_xyz:
        x, y, z = vertex_xyz[vertex]
        dual.add_vertex(vertex, x=x, y=y, z=z)

    for face in face_vertices:
        dual.add_face(face_vertices[face], fkey=face)

    if not include_boundary:
        return dual

    for boundary in mesh.faces_on_boundaries():
        for face in boundary:
            if not dual.has_vertex(face):
                x, y, z = face_centroid[face]
                dual.add_vertex(key=face, x=x, y=y, z=z)

    edge_vertex = {}
    for boundary in mesh.edges_on_boundaries():
        for u, v in boundary:
            x, y, z = mesh.edge_midpoint(u, v)
            edge_vertex[u, v] = edge_vertex[v, u] = dual.add_vertex(x=x, y=y, z=z)

    vertex_vertex = {}
    for boundary in mesh.vertices_on_boundaries():
        if boundary[0] == boundary[-1]:
            boundary = boundary[:-1]
        for vertex in boundary:
            x, y, z = mesh.vertex_coordinates(vertex)
            vertex_vertex[vertex] = dual.add_vertex(x=x, y=y, z=z)

    for boundary in mesh.vertices_on_boundaries():
        if boundary[0] == boundary[-1]:
            boundary = boundary[:-1]
        for vertex in boundary:
            vertices = [vertex_vertex[vertex]]
            nbrs = mesh.vertex_neighbors(vertex, ordered=True)[::-1]
            vertices.append(edge_vertex[vertex, nbrs[0]])
            for nbr in nbrs[:-1]:
                vertices.append(mesh.halfedge_face(vertex, nbr))
            vertices.append(edge_vertex[vertex, nbrs[-1]])
            dual.add_face(vertices[::-1])

    return dual
```
Screenshot 2022-05-03 at 15 20 42

**Comment by GeneKao** [comment]
I see, that makes a lot of sense. I will change the code accordingly.

File: CHANGELOG.md [hunk 7]
```diff
@@ -18,6 +18,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 * Added `Data.guid` to JSON serialization.
 * Added `Data.guid` to pickle state.
 * Added `Assembly.find_by_key` to locate parts by key.
+* Added `Mesh.mesh_dual` with additional option to include boundary faces.
```
Comment (tomvanmele) [inline CHANGELOG.md]: this should be under "changed" => "Changed `Mesh.mesh_dual` with option of including the boundary." or something like that...

File: src/compas/datastructures/mesh/duality.py [hunk 8]
```diff
@@ -21,20 +21,36 @@ def mesh_dual(mesh, cls=None):
     cls : Type[:class:`~compas.datastructures.Mesh`], optional
         The type of the dual mesh.
         Defaults to the type of the provided mesh object.
+    include_boundary: str, optional
```
Comment (tomvanmele) [inline src/compas/datastructures/mesh/duality.py]: type of `include_boundary` is `bool`. usual format for boolean annotations is to say what happens if the value is True.

"If True, ...".

**Review by tomvanmele — COMMENTED** [review]
almost there

**Review by tomvanmele — APPROVED** [review]
